### ISSUE TITLE
Allow defining GenericAlias-like things

### ISSUE THREAD

**Issue by adriangb** [python/cpython#103721]
# Feature or enhancement

Allow defining a GenericAlias-like thing that implements custom logic but can still interact with `typing.py` as if it was a `GenericAlias`.

# Pitch

This is useful for things that want to do runtime type checking/tracking, including Pydantic and Numpy, and has zero impact on static type checkers.

## Pydantic

```python
class Model(BaseModel, Generic[T]):
  x: int

# need to track TypeVar substitution for data validation
# requiring BaseModel.__class_getitem__ to return something
# that is not a GenericAlias
Model[int](x='abc') 

# needs to work at runtime, currently crashes w/ typing_extensions and only works on some version of CPython (including current main)
# but could be broken at any moment without a test
List[Model[T]][int]

# ideally these would work both for Pydantic's sake and for any other tools introspecting types
get_args(Model[T][int])
```

## Numpy

I'm attempting to summarize how this applies to the [discusion in  TypingSig](https://mail.python.org/archives/list/typing-sig@python.org/thread/LRDYM2MR3ZNXG6IUFTDCDJBOOVDT2U72/#WQIGF6D272DMWFJXMFXVUNVPV2Q4LKGO) but full disclosure I am not the original author or a Numpy developer.

```python
ST = TypeVar("ST")
DT = np.dtype[ST]
actual = DT[np.float64]
```

If `np.dtype[ST]` returns a "generic alias like" thing then that thing can implement `__getitem__` and return a concrete `np.Float64Dtype` or whatever when it gets substituted with `np.float64`.

## Implementation

Most of this already works in `typing.py` on CPython main. I only will need to tweak `get_args` and `get_origin`.
On the other hand, `typing_extensions` and older versions of CPython break if you do `List[MyGenericClass[T]][int]` because `List` doesn't recognize that `MyGenericClass` has generic parameters. So half other the feature request here is to add tests for existing behavior so that it is standardized and doesn't change going forward.

# Previous discussion

Discussed a PyCon 2023 typing summit, my personal impression was a generally positive reception of the features request.

### Linked PRs
* gh-103722
* gh-103838
* gh-103840

**Comment by JelleZijlstra** [comment]
I support the idea of making use cases like yours work better, and adding tests to make sure the behavior doesn't regress.

I'm not sure I like the approach in the current PR, which looks at a specific list of attributes that GenericAlias happens to have and requires all of them. Maybe code like `typing.get_origin` should simply skip the type checks and do `getattr(obj, "__origin__", None)`.

A related issue @sobolevn brought up is that this may break some code that uses `isinstance` to define generic aliases. That code is using `typing._GenericAlias`, which is supposed to be private. Does that mean we need a new public API to support what that code is trying to do?

**Comment by adriangb** [comment]
I agree that the current implementation in the PR is not ideal. I was shooting for “least amount of changes” but maybe that’s not what we need to optimize for here.

The issue that Nikita brought up makes sense. External libraries may want to not only create these things but also introspect them.

I spent some time last night experimenting and I found `types.GenericAlias`. I say “found” because it is not well documented, but it does exist (as of 3.9.0), is public, can be subclasses (as of 3.9.2). So maybe that’s all we need.

Usability of `types.GenericAlias` is however not great. For starters, it is not well documented at all. It also doesn’t allow you to implement `__init__` to add parameters, you must implement `__new__`. Some usability improvements and better documentation would go a long way.

I am also a bit confused as to why we have `typing._GenericAlias` and `types.GenericAlias`. I realize the former predates the latter, but is there a good reason to not insert `types.GenericAlias` into the bases of `typing._GenericAlias` such that libraries can just check for `types.GenericAlias`?

**Comment by JelleZijlstra** [comment]
I think the general answer is that these things weren't defined with introspectability and extensibility in mind, and they accreted over time (`typing._GenericAlias` predates `types.GenericAlias`, and we didn't think to rewrite `typing.py` when the builtin generics were introdued). We should survey the use cases we want to cover better and design a better API that remains backwards compatible.

Documenting GenericAlias and subclassing it sounds like a promising approach.

**Comment by adriangb** [comment]
That makes sense.

I think Pydantic will experiment with subclassing `types.GenericAlias` in the next few days so we’ll have some more feedback, but based on preliminary testing I think it may work as is, better documentation and QOL improvements would be a cherry on top. This will give us a good idea of what producers need.

Regarding use in `typing.py`: although I’m sure there’s some much needed refactoring, even if all we did was override `__issubclass__` to make `isinstance(List[int], types.GenericAlias)` True we might be set as far as consumers are concerned.

**Comment by JelleZijlstra** [comment]
@adriangb Do you happen to have a link to the PR/issue that made GenericAlias subclassable? I'm curious what the motivation was.

**Comment by adriangb** [comment]
Me too. I’ll look in a bit. I only found out because of a 1 liner in the docs: https://docs.python.org/3/library/types.html#types.GenericAlias

**Comment by AlexWaygood** [comment]
Note that most of the docs for `types.GenericAlias` are here rather that in `types.rst`: https://docs.python.org/3/library/stdtypes.html#generic-alias-type

(if that's confusing, then that's on us, and we should make it clearer!)

**Comment by adriangb** [comment]
Thanks, that is helpful. Good to know that `types.GenericAlias` has more than that small blurb of documentation! I do see now that the link "The type of **parameterized generics** such as list[int]" links to the right place. I guess it's on me for not clicking it, but maybe being more explicit about "here are the docs for `GenericAlias`" or moving [this section](https://docs.python.org/3/library/stdtypes.html#special-attributes-of-genericalias-objects) to the `types.py` docs would help.

If we do end up deciding that `types.GenericAlias` is the solution for this use case it might also be nice to document that. Could we put an example in the docs showing how a hypothetical library could customize `types.GenericAlias` by subclassing it?

As far as the general suitability of `types.GenericAlias` for this use case, my preliminary experiments showed promise. The main thing that was a bit clunky was having to override `__new__`. Are there some easy changes we can make on the C side so that this can be overridden more like a regular Python class?
I, unfortunately, won't be able to go all the way with Pydantic because our current machinery is very entrenched and we want to make a new major release, it would be too much refactoring right now. We will definitely pick that back this work back up in a couple of weeks, I think if we can make it work in Pydantic it would be a very good sign for producers of these things.

Regarding consumers, I think the path forward is to decide if (1) it would even make sense to refactor `typing._GenericAlias` to be based on `types.GenericAlias`, and if so we could (2) somehow insert `types.GenericAlias` into `typing._GenericAlias` (override `__issubclass__`, monkey patch `__bases__` or actually inherit from it) so that consumers can do `isinstance(thing, types.GenericAlias)` instead of `isinstance(thing, typing._GenericAlias)`. As far as I can tell all consumers (at least Pydantic and Hypothesis) only need to know if it is a GenericAlias-like to get `__parameters__`, `__args__` and `__origin__`.

**Comment by AlexWaygood** [comment]
> @adriangb Do you happen to have a link to the PR/issue that made GenericAlias subclassable? I'm curious what the motivation was.

@JelleZijlstra — it was made subclassable as part of the changes made in https://bugs.python.org/issue42195. We had to make GenericAlias subclassable so that we could subclass it in `_collections_abc.py` so that we could customise the way parameterising `collections.abc.Callable` worked.

https://github.com/python/cpython/blob/d8627999d85cc5b000dbe17180250d919f8510ad/Lib/_collections_abc.py#L442

Not a use case all that different to what pydantic wants to do!
